File: samples/CameraAccess/CameraAccess/OpenClaw/ToolCallModels.swift
import Foundation

// MARK: - Gemini Tool Call (parsed from server JSON)

struct GeminiFunctionCall {
  let id: String
  let name: String
  let args: [String: Any]
}

struct GeminiToolCall {
  let functionCalls: [GeminiFunctionCall]

  init?(json: [String: Any]) {
    // Original Gemini format
    if let toolCall = json["toolCall"] as? [String: Any],
       let calls = toolCall["functionCalls"] as? [[String: Any]] {
      self.functionCalls = calls.compactMap { call in
        guard let id = call["id"] as? String,
              let name = call["name"] as? String else { return nil }
        let args = call["args"] as? [String: Any] ?? [:]
        return GeminiFunctionCall(id: id, name: name, args: args)
      }
      return
    }

    // Doubao / OpenAI format wrapper (provided via GeminiLiveService mapping)
    if let calls = json["functionCalls"] as? [[String: Any]] {
      self.functionCalls = calls.compactMap { call in
        // Doubao format: { "name": "execute", "arguments": "{\"task\":\"...\"}" }
        guard let name = call["name"] as? String else { return nil }
        
        // id might not be present or named differently in some formats, fallback to UUID if needed by OpenClaw
        let id = call["call_id"] as? String ?? call["id"] as? String ?? UUID().uuidString
        
        var args: [String: Any] = [:]
        if let argsString = call["arguments"] as? String,
           let data = argsString.data(using: .utf8),
           let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
          args = parsed
        } else if let argsDict = call["arguments"] as? [String: Any] {
          args = argsDict
        }

        return GeminiFunctionCall(id: id, name: name, args: args)
      }
      return
    }

    return nil
  }
}

// MARK: - Gemini Tool Call Cancellation

struct GeminiToolCallCancellation {
  let ids: [String]

  init?(json: [String: Any]) {
    guard let cancellation = json["toolCallCancellation"] as? [String: Any],
          let ids = cancellation["ids"] as? [String] else {
      return nil
    }
    self.ids = ids
  }
}

// MARK: - Tool Result

enum ToolResult {
  case success(String)
  case failure(String)

  var responseValue: [String: Any] {
    switch self {
    case .success(let result):
      return ["result": result]
    case .failure(let error):
      return ["error": error]
    }
  }
}

// MARK: - Tool Call Status (for UI)

enum ToolCallStatus: Equatable {
  case idle
  case executing(String)
  case completed(String)
  case failed(String, String)
  case cancelled(String)

  var displayText: String {
    switch self {
    case .idle: return ""
    case .executing(let name): return "Running: \(name)..."
    case .completed(let name): return "Done: \(name)"
    case .failed(let name, let err): return "Failed: \(name) - \(err)"
    case .cancelled(let name): return "Cancelled: \(name)"
    }
  }

  var isActive: Bool {
    if case .executing = self { return true }
    return false
  }
}

// MARK: - Tool Declarations (for Gemini setup message)

enum ToolDeclarations {

  static func allDeclarations() -> [[String: Any]] {
    return [execute]
  }

  static let execute: [String: Any] = [
    "name": "execute",
    "description": "Your only way to take action. You have no memory, storage, or ability to do anything on your own -- use this tool for everything: sending messages, searching the web, adding to lists, setting reminders, creating notes, research, drafts, scheduling, smart home control, app interactions, or any request that goes beyond answering a question. When in doubt, use this tool.",
    "parameters": [
      "type": "object",
      "properties": [
        "task": [
          "type": "string",
          "description": "Clear, detailed description of what to do. Include all relevant context: names, content, platforms, quantities, etc."
        ]
      ],
      "required": ["task"]
    ]
  ]
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/openclaw/ToolCallModels.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw

import org.json.JSONArray
import org.json.JSONObject

// Gemini Tool Call (parsed from server JSON)

data class GeminiFunctionCall(
    val id: String,
    val name: String,
    val args: Map<String, Any?>
)

data class GeminiToolCall(
    val functionCalls: List<GeminiFunctionCall>
) {
    companion object {
        fun fromJSON(json: JSONObject): GeminiToolCall? {
            // Priority 1: Original Gemini format
            val toolCall = json.optJSONObject("toolCall")
            if (toolCall != null) {
                val calls = toolCall.optJSONArray("functionCalls") ?: return null
                val functionCalls = mutableListOf<GeminiFunctionCall>()
                for (i in 0 until calls.length()) {
                    val call = calls.getJSONObject(i)
                    val id = call.optString("id", "")
                    val name = call.optString("name", "")
                    if (id.isEmpty() || name.isEmpty()) continue
                    val argsObj = call.optJSONObject("args")
                    val args = mutableMapOf<String, Any?>()
                    if (argsObj != null) {
                        for (key in argsObj.keys()) {
                            args[key] = argsObj.opt(key)
                        }
                    }
                    functionCalls.add(GeminiFunctionCall(id, name, args))
                }
                return if (functionCalls.isNotEmpty()) GeminiToolCall(functionCalls) else null
            }

            // Priority 2: Doubao / OpenAI format wrapper
            val doubaoCalls = json.optJSONArray("functionCalls")
            if (doubaoCalls != null) {
                val functionCalls = mutableListOf<GeminiFunctionCall>()
                for (i in 0 until doubaoCalls.length()) {
                    val call = doubaoCalls.getJSONObject(i)
                    val name = call.optString("name", "")
                    if (name.isEmpty()) continue
                    
                    val id = call.optString("call_id", call.optString("id", java.util.UUID.randomUUID().toString()))
                    val args = mutableMapOf<String, Any?>()
                    
                    // Doubao sets "arguments" as a stringified JSON object
                    val argsStr = call.optString("arguments", null)
                    if (argsStr != null) {
                        try {
                            val parsedArgs = JSONObject(argsStr)
                            for (key in parsedArgs.keys()) {
                                args[key] = parsedArgs.opt(key)
                            }
                        } catch (e: Exception) {
                            // Ignored 
                        }
                    } else {
                        // fallback if provided as raw object
                        val argsObj = call.optJSONObject("arguments")
                        if (argsObj != null) {
                            for (key in argsObj.keys()) {
                                args[key] = argsObj.opt(key)
                            }
                        }
                    }
                    functionCalls.add(GeminiFunctionCall(id, name, args))
                }
                return if (functionCalls.isNotEmpty()) GeminiToolCall(functionCalls) else null
            }
            
            return null
        }
    }
}

// Gemini Tool Call Cancellation

data class GeminiToolCallCancellation(
    val ids: List<String>
) {
    companion object {
        fun fromJSON(json: JSONObject): GeminiToolCallCancellation? {
            val cancellation = json.optJSONObject("toolCallCancellation") ?: return null
            val idsArray = cancellation.optJSONArray("ids") ?: return null
            val ids = mutableListOf<String>()
            for (i in 0 until idsArray.length()) {
                ids.add(idsArray.getString(i))
            }
            return if (ids.isNotEmpty()) GeminiToolCallCancellation(ids) else null
        }
    }
}

// Tool Result

sealed class ToolResult {
    data class Success(val result: String) : ToolResult()
    data class Failure(val error: String) : ToolResult()

    fun toJSON(): JSONObject = when (this) {
        is Success -> JSONObject().put("result", result)
        is Failure -> JSONObject().put("error", error)
    }
}

// Tool Call Status (for UI)

sealed class ToolCallStatus {
    data object Idle : ToolCallStatus()
    data class Executing(val name: String) : ToolCallStatus()
    data class Completed(val name: String) : ToolCallStatus()
    data class Failed(val name: String, val error: String) : ToolCallStatus()
    data class Cancelled(val name: String) : ToolCallStatus()

    val displayText: String
        get() = when (this) {
            is Idle -> ""
            is Executing -> "Running: $name..."
            is Completed -> "Done: $name"
            is Failed -> "Failed: $name - $error"
            is Cancelled -> "Cancelled: $name"
        }

    val isActive: Boolean
        get() = this is Executing
}

// OpenClaw Connection State

sealed class OpenClawConnectionState {
    data object NotConfigured : OpenClawConnectionState()
    data object Checking : OpenClawConnectionState()
    data object Connected : OpenClawConnectionState()
    data class Unreachable(val message: String) : OpenClawConnectionState()
}

// Tool Declarations (for Gemini setup message)

object ToolDeclarations {
    fun allDeclarationsJSON(): JSONArray {
        return JSONArray().put(executeJSON())
    }

    private fun executeJSON(): JSONObject {
        return JSONObject().apply {
            put("name", "execute")
            put("description", "Your only way to take action. You have no memory, storage, or ability to do anything on your own -- use this tool for everything: sending messages, searching the web, adding to lists, setting reminders, creating notes, research, drafts, scheduling, smart home control, app interactions, or any request that goes beyond answering a question. When in doubt, use this tool.")
            put("parameters", JSONObject().apply {
                put("type", "object")
                put("properties", JSONObject().apply {
                    put("task", JSONObject().apply {
                        put("type", "string")
                        put("description", "Clear, detailed description of what to do. Include all relevant context: names, content, platforms, quantities, etc.")
                    })
                })
                put("required", JSONArray().put("task"))
            })
        }
    }
}
